Merge allocator and store finder; remove store finder.

spencerkimball opened the pull request:
Implemented two methods on allocator: Allocate() and Rebalance().
Rebalance() is what we'll use most of the time and always chooses
a random store which will aid in rebalancing. Allocate() chooses
from all stores with a bias towards ones with more capacity. This
will be used when recovering from dead stores/nodes.

spencerkimball committed:
Merge allocator and store finder; remove store finder.

Implemented two methods on allocator: Allocate() and Rebalance().
Rebalance() is what we'll use most of the time and always chooses
a random store which will aid in rebalancing. Allocate() chooses
from all stores with a bias towards ones with more capacity. This
will be used when recovering from dead stores/nodes.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, capacity *stat) bool {
+		// Use counts instead of capacities if the cluster has mean capacity
```
It'd be nice to do this without an arbitrary threshold, so that `filter` is not discontinuous

spencerkimball replied:
I don't think a fuzzy threshold (presumably with different probabilities of selection) would add very much utility. It would add complexity. Perhaps you had something else in mind?

bdarnell replied:
I think we could get rid of rebalanceFromMean and replace it with a dynamically-computed value based on the average size of a range (i.e. a store must be above/below the mean by at least the size of a range to be a target, whether one range is 10% or 1% of the store). If we have that I'm not sure if the special case for small clusters is necessary.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+func (a *allocator) selectRandom(count int, required proto.Attributes, existing []proto.Replica,
+	filter func(*proto.StoreDescriptor, *stat, *stat) bool) ([]*proto.StoreDescriptor, error) {
+	var descs []*proto.StoreDescriptor
+	sl := a.findStores(required)
```
`sl`? i'd like to buy a vowel

spencerkimball replied:
`storeList`.... I'm going to rename `findStores` to `getStoreList` to make this clearer.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+}
+
+// Update updates count and capacity stats using the specified descriptor.
+func (sl *storeList) Update(s *proto.StoreDescriptor) {
```
nit: can the name and comment reflect that this method holds on to the descriptor? also, since `storeList` holds on to the descriptor (indefinitely?), the fact that `count` and `capacity` are `stat`s is an implementation detail, and should not be referenced in the doc comment

spencerkimball replied:
Changed name to Add(). I made the comment a little less specific, but I still do want to mention that adding the store descriptor updates statistics for the list.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
-		capacitySeen += c.Capacity.PercentAvail()
-		if capacitySeen >= targetCapacity {
-			return c, nil
+	if a.deterministic {
```
what behaviour does this alter?

spencerkimball replied:
It sorts the keys from the capacityKeys set so that the random picking permutation is deterministic assuming a deterministic seed to the rng. The order which sets are traversed is based on something else.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+	defer a.Unlock()
+	sl := a.findStores(*s.CombinedAttrs())
+
+	if sl.capacity.mean > maxAvailCapacityThreshold {
```
this logic is nearly duplicated in `Rebalance`

spencerkimball replied:
Nearly, but not near enough. I had it originally as a separate method but required passing a flag in which basically resulted in an if statement inside the method, bifurcating the code paths there. This is actually the cleaner approach.

petermattis commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, contentsChanged bool) {
```
`contentsChanged` is unused. Can it be removed?

tamird replied:
this method is used as a callback, so its required signature is defined elsewhere. I'd +1 a change to `_ bool` though.

spencerkimball replied:
Changed to `_`

vivekmenezes commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+	maxAvailCapacityThreshold = 0.98
+	// rebalancePctFromMean: if the capacity or range count of a store
+	// is within rebalancePctFromMean of the mean, it is considered a
+	// viable target for rebalancing.
```
s/for rebalancing/to rebalance to/

spencerkimball replied:
Done.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
```
this is confusing: select 3 random stores (why 3?), and then pick the one with the highest percentage capacity; why do we want randomness here? is it just to protect against thundering herds? if so, can we introduce some information into these structures that tracks in-progress rebalances?

Also, it seems like we're going to want to consider absolute space as well as percentage when making these decisions; are there plans to implement that?

spencerkimball replied:
Re: absolute space...one step at a time. I actually doubt absolute space will be something we ever get around to worrying about. The next item will be iops load rebalancing.

The explanation for 3 random stores is in the struct doc for `allocator`. It's an attempt to be very random with placement, but provide a bias towards balancing.

Thing is, based on how our clusters grow by splitting in situ, the straight up allocation is relatively rare and comes into play only when there's a machine or disk death. Otherwise, we're always rebalancing.

petermattis commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+// the required attributes, do not include any of the existing
+// replicas, and return true when passed to the provided filter.
+// If the supplied filter is nil, it is ignored.
+func (a *allocator) selectRandom(count int, required proto.Attributes, existing []proto.Replica,
```
`s/selectRandom/selectRandomLocked/g`?

spencerkimball replied:
So far I only use that naming when we have both a Locked version and a non-Locked version of the same method.

There are countless places in the code where we don't follow this naming convention you're suggesting. Ultimately, it would be nice to have an AssertHeld on the mutex, but failing that, I don't want to switch our naming convention arbitrarily, which is what this feels like.

vivekmenezes commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+// from amongst stores which fall within an acceptable multiple
+// of standard deviations from either the capacity or range count
+// means.
+func (a *allocator) Rebalance(required proto.Attributes, existing []proto.Replica) (
```
s/Rebalance/RebalanceTarget ?

spencerkimball replied:
This is symmetric with Allocate. I'll rename both with Target suffix.

petermattis commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+// TODO(embark, spencer): consider using a reverse index map from
+// Attr->stores, for efficiency. Ensure that entries in this map still
+// have an opportunity to be garbage collected.
+func (a *allocator) findStores(required proto.Attributes) *storeList {
```
`s/findStores/findStoresLocked/g`?

spencerkimball replied:
See above.

petermattis commented on storage/allocator_test.go:
```diff
@@ -293,3 +289,438 @@ func TestExistingReplica(t *testing.T) {
+	// fc d2 e8 f5 fd e7 da e7 d6 f6 e2 d3 ff f8 f9 e7 e0 f2 f9 d0
+	// ee d8 df f2 f8 e4 dc e0 d5 f2 dd d2 ff f2 f3 e7 de e5 f9 d3
+	// ee e0 e4 f0 ff ec db e4 d6 f4 dc d7 fd ef ee eb df e9 fd cf
+	// f1 e4 e7 ec fa ee e0 e9 d5 fa dd dd ff f0 f2 ef e3 f0 fc d1
```
That's impenetrable. Perhaps output as 3 character percentages.

spencerkimball replied:
Fine, changed to 3 digit values from 0-999.

petermattis replied:
I was thinking 3 digit percentages from 0-100. Or if you know you'll never reach 100% you could do 0-99.

petermattis commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
-// availability of servers is gleaned from the gossip network.
+const (
+	// minAvailCapacityThreshold: if the capacity of a store descriptor
+	// falls below this percentage, it will never be used as an
```
Nit: this is a fraction, not a percentage. I might rename percent/percentage to fraction everywhere to make this clearer.

spencerkimball replied:
Changed all the names I could get my hands on and fixed comments.

BramGruneir commented on storage/allocator.go:
```diff
@@ -13,75 +13,291 @@
+	// is within rebalancePctFromMean of the mean, it is considered a
+	// viable target for rebalancing.
+	rebalancePctFromMean = 0.025 // 2.5%
+)
```
How did you come up with these values?  I"m sure they're good for now, but perhaps we should pass them in so we can test under load with different thresholds?

spencerkimball replied:
Pulled 'em straight out of my ass.

We can parameterize them if/when we need to. I'll be surprised if we change these anytime soon, except to tweak them when we watch behavior in the real world. I don't particularly want to expose any knobs unless we find a solid use case for different modes of operation.

spencerkimball committed:
Address feedback

BramGruneir commented on storage/allocator.go:
```diff
@@ -13,75 +13,289 @@
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, used *stat) bool {
```
There's no need for this to be an anonymous function, why not pull it out and give it a full description?  Call it rebalanceFilter or something like that.

spencerkimball replied:
Feels right to me. It's so integral to how we're using `selectRandom` from this method. When I was trying to share this code with `shouldRebalance()`, I broke it out. But when the sharing didn't work, I ended up wanting to keep it local as part of this method.

bdarnell commented on storage/allocator.go:
```diff
@@ -13,75 +13,289 @@
+	if sl.used.mean < minFractionUsedThreshold {
+		return float64(s.Capacity.RangeCount) > sl.count.mean
+	}
+	return s.Capacity.FractionUsed() > sl.used.mean
```
This should have some margin for error (maybe add rebalanceFromMean, just like we subtract it when looking for targets) instead of trying to rebalance every range that is above the mean. The threshold should probably be derived from the number of ranges per store - when there are few ranges, moving one of them could cause a swing of more than 2.5%.

spencerkimball replied:
Well there is a margin of error--it's applied on the other side. That is, it's applied when selecting a list of possible rebalance targets. If that list is empty, the rebalance will not happen.

The sort of flapping you're worried about with the 2.5% `rebalanceFromMean` constant won't come into play. That's part of why we have the `minFractionUsedThreshold`. This says that you need at least 5% mean bytes used across all stores with matching attributes in order to make decisions on allocation targets according to fraction used. Of course with tiny byte values you can still easily cause flapping, this won't be a problem in practice with (I imagine) 10s to 1000s of GB per store.

bdarnell replied:
I'd rather be conservative and apply the margin of error on both sides. It's more important to prevent an overactive rebalancer than to come as close as possible to perfectly balanced.

We need some sort of margin of error for the count-based case too: otherwise a four-node cluster with a single range (replicated 3x) would constantly rebalance until it hits `minFractionUsedThreshold`, since the three nodes with a range would be above the mean and the empty node would be below the mean. 

We should probably add the range to be moved to the target store's stats in RebalanceTarget before comparing to the mean, so we don't make a move that will immediately cause rebalancing pressure in the other direction.

spencerkimball replied:
The special case for small clusters was actually motivated by something else. I'm planning for when people want to start up a cluster and manually split it in order to do parallel data importing. I want those small, manually split ranges to rebalance. There are other ways to do this, but this seemed like a reasonable way to get that behavior in the short run. Eventually, manual splits (and load-motivated splits) will want a hint for their ranges to specifically rebalance.

I'll add a margin of error for the count-based case.

spencerkimball committed:
fix failing unittest

bdarnell commented:
LGTM, we can work out the details of the algorithm (and how we prevent flapping, etc) in later changes.

spencerkimball committed:
Added an error margin for range count based rebalance decisions.

Changed RebalanceTarget() to use AllocateTarget() more exactly. That is,
we now choose three random targets from all available stores which match
the attributes, and find the best result which matches rebalancing criteria.

The difference is that now we will be less likely to find a target
successfully in the event that there are enough potential targets which
aren't rebalanceable. However, this is actually OK, because the ratio
of balancers to balancees for the particular class of store evens the
odds so we expect balancings to continue at the correct pace without
overcrowding any particular target or targets.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,313 @@
-		// TODO(bdarnell): use a real random seed.
-		rand: *rand.New(rand.NewSource(0)),
+// newAllocator creates a new allocator for the store specified by
+// storeID.
```
where is storeID?

spencerkimball replied:
Done.

tamird commented on storage/allocator.go:
```diff
@@ -13,75 +13,313 @@
+// getUsedNodes returns a set of node IDs which are already being used
+// to store replicas.
+func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
 	// Get a set of current nodes -- we never want to allocate on an existing node.
```
nit: remove this comment

tamird commented on storage/allocator_test.go:
```diff
@@ -46,141 +53,135 @@ var multiDCConfig = proto.ZoneConfig{
-func filterStores(a proto.Attributes, stores []*proto.StoreDescriptor) ([]*proto.StoreDescriptor, error) {
-	var filtered []*proto.StoreDescriptor
+func gossipStores(g *gossip.Gossip, stores []*proto.StoreDescriptor, t *testing.T) {
+	var wg sync.WaitGroup
```
this machinery (wrapping the iteration of `stores`) looks duplicated. can it be extracted?

spencerkimball replied:
So far it's duplicated in two different packages. Not quite ready to extract it.

tamird commented on storage/allocator_test.go:
```diff
@@ -46,141 +53,135 @@ var multiDCConfig = proto.ZoneConfig{
+	s, _, stopper := createTestStore(t)
+	defer stopper.Stop()
+	gossipStores(s.Gossip(), singleStore, t)
+	result, err := s._allocator.AllocateTarget(simpleZoneConfig.ReplicaAttrs[0], []proto.Replica{})
```
nit: s/_allocator/allocator()/

spencerkimball replied:
Done.

tamird commented on storage/allocator_test.go:
```diff
@@ -189,13 +190,11 @@ func TestSimpleRetrieval(t *testing.T) {
-	result, err := a.allocate(simpleZoneConfig.ReplicaAttrs[0], []proto.Replica{})
+	s, _, stopper := createTestStore(t)
+	defer stopper.Stop()
+	result, err := s._allocator.AllocateTarget(simpleZoneConfig.ReplicaAttrs[0], []proto.Replica{})
```
nit: s/_allocator/allocator()/

tamird replied:
Oh there are a bunch of instances of this

spencerkimball replied:
Changed.

tamird commented on storage/allocator_test.go:
```diff
@@ -273,16 +270,15 @@ func TestTwoDatacenters(t *testing.T) {
-		rand:        *rand.New(rand.NewSource(0)),
-	}
-	result, err := a.allocate(multiDisksConfig.ReplicaAttrs[1], []proto.Replica{
+	s, _, stopper := createTestStore(t)
```
my kingdom for some macros
